Fretboard: JSONException

From Focus.

```
Caused by: org.json.JSONException: 
  at org.json.JSONTokener.syntaxError (JSONTokener.java:449)
  at org.json.JSONTokener.nextValue (JSONTokener.java:97)
  at org.json.JSONObject.<init> (JSONObject.java:156)
  at org.json.JSONObject.<init> (JSONObject.java:173)
  at mozilla.components.service.fretboard.storage.flatfile.ExperimentsSerializer.fromJson (ExperimentsSerializer.kt:47)
  at mozilla.components.service.fretboard.storage.flatfile.FlatFileExperimentStorage.retrieve (FlatFileExperimentStorage.kt:25)
  at mozilla.components.service.fretboard.Fretboard.loadExperiments (Fretboard.kt:39)
  at org.mozilla.focus.FocusApplication.loadExperiments (FocusApplication.kt:97)
  at org.mozilla.focus.FocusApplication$onCreate$1.doResume (FocusApplication.kt:60)
  at kotlin.coroutines.experimental.jvm.internal.CoroutineImpl.resume (CoroutineImpl.kt:42)
  at kotlinx.coroutines.experimental.DispatchedTask$DefaultImpls.run (Dispatched.kt:162)
  at kotlinx.coroutines.experimental.DispatchedContinuation.run (Dispatched.kt:26)
  at kotlinx.coroutines.experimental.EventLoopBase.processNextEvent (EventLoop.kt:148)
  at kotlinx.coroutines.experimental.BlockingCoroutine.joinBlocking (Builders.kt:82)
  at kotlinx.coroutines.experimental.BuildersKt__BuildersKt.runBlocking (Builders.kt:58)
  at kotlinx.coroutines.experimental.BuildersKt.runBlocking (Unknown Source)
  at kotlinx.coroutines.experimental.BuildersKt__BuildersKt.runBlocking$default (Builders.kt:48)
  at kotlinx.coroutines.experimental.BuildersKt.runBlocking$default (Unknown Source)
  at org.mozilla.focus.FocusApplication.onCreate (FocusApplication.kt:59)
  at android.app.Instrumentation.callApplicationOnCreate (Instrumentation.java:1014)
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:4800)
```

--- a/components/service/fretboard/src/main/java/mozilla/components/service/fretboard/storage/flatfile/ExperimentsSerializer.kt
+++ b/components/service/fretboard/src/main/java/mozilla/components/service/fretboard/storage/flatfile/ExperimentsSerializer.kt
@@ -8,6 +8,7 @@ import mozilla.components.service.fretboard.Experiment
 import mozilla.components.service.fretboard.JSONExperimentParser
 import mozilla.components.service.fretboard.ExperimentsSnapshot
 import org.json.JSONArray
+import org.json.JSONException
 import org.json.JSONObject
 
 /**
@@ -43,6 +44,7 @@ internal class ExperimentsSerializer {
      *
      * @return experiment snapshot with the parsed list of experiments
      */
+    @Throws(JSONException::class)
     fun fromJson(json: String): ExperimentsSnapshot {
         val experimentsJson = JSONObject(json)
         val experimentsJsonArray = experimentsJson.getJSONArray(EXPERIMENTS_KEY)

--- a/components/service/fretboard/src/main/java/mozilla/components/service/fretboard/storage/flatfile/FlatFileExperimentStorage.kt
+++ b/components/service/fretboard/src/main/java/mozilla/components/service/fretboard/storage/flatfile/FlatFileExperimentStorage.kt
@@ -7,6 +7,7 @@ package mozilla.components.service.fretboard.storage.flatfile
 import android.util.AtomicFile
 import mozilla.components.service.fretboard.ExperimentStorage
 import mozilla.components.service.fretboard.ExperimentsSnapshot
+import org.json.JSONException
 import java.io.FileNotFoundException
 import java.io.File
 import java.io.IOException
@@ -24,6 +25,10 @@ class FlatFileExperimentStorage(file: File) : ExperimentStorage {
             val experimentsJson = String(atomicFile.readFully())
             ExperimentsSerializer().fromJson(experimentsJson)
         } catch (e: FileNotFoundException) {
+            ExperimentsSnapshot(listOf(), null)
+        } catch (e: JSONException) {
+            // The JSON we read from disk is corrupt. There's nothing we can do here and therefore
+            // we just continue as if the file wouldn't exist.
             ExperimentsSnapshot(listOf(), null)
         }
     }

--- a/components/service/fretboard/src/test/java/mozilla/components/service/fretboard/FretboardTest.kt
+++ b/components/service/fretboard/src/test/java/mozilla/components/service/fretboard/FretboardTest.kt
@@ -10,6 +10,7 @@ import android.content.pm.PackageInfo
 import android.content.pm.PackageManager
 import kotlinx.coroutines.experimental.CommonPool
 import kotlinx.coroutines.experimental.runBlocking
+import mozilla.components.service.fretboard.storage.flatfile.FlatFileExperimentStorage
 import mozilla.components.support.test.any
 import org.junit.Assert.assertEquals
 import org.junit.Assert.assertFalse
@@ -25,6 +26,7 @@ import org.mockito.Mockito.times
 import org.mockito.Mockito.verify
 import org.robolectric.RobolectricTestRunner
 import org.robolectric.RuntimeEnvironment
+import java.io.File
 import kotlin.reflect.full.functions
 import kotlin.reflect.jvm.isAccessible
 
@@ -560,5 +562,20 @@ class FretboardTest {
                 .forEach {
                     assertTrue(it.value in 350..650)
                 }
+    }
+
+    @Test
+    fun testLoadingCorruptJSON() {
+        val experimentSource = mock(ExperimentSource::class.java)
+
+        val file = File(RuntimeEnvironment.application.filesDir, "corrupt-experiments.json")
+        file.writer().use {
+            it.write("""{"experiment":[""")
+        }
+
+        val experimentStorage = FlatFileExperimentStorage(file)
+
+        val fretboard = Fretboard(experimentSource, experimentStorage)
+        fretboard.loadExperiments() // Should not throw
     }
 }

--- a/components/service/fretboard/src/test/java/mozilla/components/service/fretboard/storage/flatfile/FlatFileExperimentStorageTest.kt
+++ b/components/service/fretboard/src/test/java/mozilla/components/service/fretboard/storage/flatfile/FlatFileExperimentStorageTest.kt
@@ -125,4 +125,17 @@ class FlatFileExperimentStorageTest {
         assertEquals(0, experimentsResult.experiments.size)
         assertNull(experimentsResult.lastModified)
     }
+
+    @Test
+    fun testReadingCorruptJSON() {
+        val file = File(RuntimeEnvironment.application.filesDir, "corrupt-experiments.json")
+        file.writer().use {
+            it.write("""{"experiment":[""")
+        }
+
+        val snapshot = FlatFileExperimentStorage(file).retrieve()
+
+        assertNull(snapshot.lastModified)
+        assertEquals(0, snapshot.experiments.size)
+    }
 }

